File: app/src/main/java/com/stathis/elmepaunivapp/recyclerview/DepMembersAdapter.java
package com.stathis.elmepaunivapp.recyclerview;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.stathis.elmepaunivapp.R;
import com.stathis.elmepaunivapp.listeners.OnItemClickListener;
import com.stathis.elmepaunivapp.models.DepMembers;

import java.util.List;

public class DepMembersAdapter extends RecyclerView.Adapter<DepMembersViewHolder> {

    private List<DepMembers> depMembers;
    private OnItemClickListener onItemClickListener;

    public DepMembersAdapter(List<com.stathis.elmepaunivapp.models.DepMembers> depMembers, OnItemClickListener listener) {
        this.depMembers = depMembers;
        onItemClickListener = listener;
    }

    @NonNull
    @Override
    public DepMembersViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.dep_members_item_row, parent, false);
        return new DepMembersViewHolder(view,onItemClickListener);
    }

    @Override
    public void onBindViewHolder(@NonNull DepMembersViewHolder holder, int position) {
        holder.present(depMembers.get(position));
    }

    @Override
    public int getItemCount() {
        return depMembers.size();
    }
}

File: feature/department/src/main/java/com/stathis/department/department/adapter/DepMembersAdapter.kt
package com.stathis.department.department.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.ListAdapter
import com.stathis.common.base.BaseDiffUtil
import com.stathis.common.base.BaseViewHolder
import com.stathis.common.databinding.HolderEmptyViewBinding
import com.stathis.department.BR
import com.stathis.department.R
import com.stathis.department.databinding.HolderDepmemberItemBinding
import com.stathis.department.databinding.HolderShimmerDepmemberItemBinding
import com.stathis.model.UiModel
import com.stathis.model.general.ShimmerItem
import com.stathis.model.personnel.Person

class DepMembersAdapter : ListAdapter<com.stathis.model.UiModel, DepMembersViewHolder>(BaseDiffUtil<UiModel>()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DepMembersViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = when (viewType) {
            R.layout.holder_depmember_item -> {
                HolderDepmemberItemBinding.inflate(inflater, parent, false)
            }

            R.layout.holder_shimmer_depmember_item -> {
                HolderShimmerDepmemberItemBinding.inflate(inflater, parent, false)
            }

            else -> HolderEmptyViewBinding.inflate(inflater, parent, false)
        }
        return DepMembersViewHolder(view)
    }

    override fun onBindViewHolder(holder: DepMembersViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    override fun getItemViewType(position: Int) = when (getItem(position)) {
        is Person -> R.layout.holder_depmember_item
        is ShimmerItem -> R.layout.holder_shimmer_depmember_item
        else -> com.stathis.common.R.layout.holder_empty_view
    }
}

class DepMembersViewHolder(
    private val binding: ViewDataBinding
) : BaseViewHolder(binding) {

    override fun bind(data: com.stathis.model.UiModel) {
        binding.setVariable(BR.model, data)
    }
}